google/cloud/bigtable/read_rows_query.py

```diff
@@ -13,36 +13,158 @@
+          - ValueError if new_limit is < 0
+        """
+        if new_limit is not None and new_limit < 0:
+            raise ValueError("limit must be >= 0")
+        self._limit = new_limit
+        return self
+
+    def set_filter(
+        self, row_filter: RowFilter | dict[str, Any] | None
+    ) -> ReadRowsQuery:
+        """
+        Set a RowFilter to apply to this query
+
+        Args:
+          - row_filter: a RowFilter to apply to this query
+              Can be a RowFilter object or a dict representation
+        Returns:
+          - a reference to this query for chaining
+        """
+        if not (
+            isinstance(row_filter, dict)
+            or isinstance(row_filter, RowFilter)
+            or row_filter is None
+        ):
+            raise ValueError("row_filter must be a RowFilter or dict")
+        self._filter = row_filter
+        return self
+
+    def add_rows(self, row_keys: list[str | bytes] | str | bytes) -> ReadRowsQuery:
+        """
+        Add a list of row keys to this query
+
+        Args:
+          - row_keys: a list of row keys to add to this query
+        Returns:
+          - a reference to this query for chaining
+        Raises:
+          - ValueError if an input is not a string or bytes
+        """
+        if not isinstance(row_keys, list):
+            row_keys = [row_keys]
+        update_set = set()
+        for k in row_keys:
+            if isinstance(k, str):
+                k = k.encode()
+            elif not isinstance(k, bytes):
+                raise ValueError("row_keys must be strings or bytes")
+            update_set.add(k)
+        self.row_keys.update(update_set)
+        return self
 
     def add_range(
-        self, start_key: str | bytes | None = None, end_key: str | bytes | None = None
+        self,
+        start_key: str | bytes | None = None,
+        end_key: str | bytes | None = None,
+        start_is_inclusive: bool | None = None,
+        end_is_inclusive: bool | None = None,
     ) -> ReadRowsQuery:
-        raise NotImplementedError
+        """
+        Add a range of row keys to this query.
+
+        Args:
+          - start_key: the start of the range
+              if None, start_key is interpreted as the empty string, inclusive
+          - end_key: the end of the range
+              if None, end_key is interpreted as the infinite row key, exclusive
+          - start_is_inclusive: if True, the start key is included in the range
+              defaults to True if None. Must not be included if start_key is None
+          - end_is_inclusive: if True, the end key is included in the range
+              defaults to False if None. Must not be included if end_key is None
+        """
+        # check for invalid combinations of arguments
+        if start_is_inclusive is None:
+            start_is_inclusive = True
+        elif start_key is None:
+            raise ValueError("start_is_inclusive must not be set without start_key")
```

Comment:
Maybe rephrase it to 
```suggestion
            raise ValueError("start_is_inclusive must be set with start_key")
```
so it's easier to understand :)